### PR TITLE
Set the core partition set size properly for batch ingestion with dynamic partitioning

### LINKED ISSUE #9972: Segments created by an overwriting task with dynamic partitioning doesn't atomically overshadow old segments in the overlapping time chunk
### Affected Version

All versions

### Description

The overwriting task such as the compaction task with the dynamic partitioning creates segments with `NumberedShardSpec`. This `NumberedShardSpec` has the `partitions` field indicating the core partition size. However, it is always set to 0 for batch tasks. Since the core partition size is 0, the segment update in the broker/coordinator server view is not guaranteed to be done atomically even in the overlapping time chunk. Instead, a new segment will overshadow immediately when a first historical announces a new segment. As a result, some queries can return less results than normal until all new segments are available in historicals.

This issue doesn't exist with hash or range partitioning.

### PR DESCRIPTION
**Pull request by jihoonson**
Fixes #9972.

### Description

This PR add a new shardSpec to fix #9972 which is used in batch ingestion with dynamic partitioning. The new shardSpec contains only the partitionId and is converted to the `NumberedShardSpec` with the proper core set size when the segment is published.

<hr>

This PR has:
- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [x] added documentation for new or modified features or behaviors.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/licenses.yaml)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

<hr>

##### Key changed/added classes in this PR
 * `BuildingNumberedShardSpec`
 * `SegmentPublisherHelper`

### CODE REVIEW

File: core/src/main/java/org/apache/druid/timeline/partition/BuildingNumberedShardSpec.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+ * segments will be created per time chunk. However, in batch ingestion with time chunk locking, the core partition
+ * set is the set of segments created by an initial task or an overwriting task. Since the core partition set is
+ * determined when the task publishes segments at the end, the task postpones creating proper NumberedShardSpec
+ * until the end.
+ *
+ * This shardSpec is used for such use case. A non-appending batch task can use this shardSpec until it publishes
+ * segments at last. When it publishes segments, it should convert the shardSpec of those segments to NumberedShardSpec.
+ * See {@code SegmentPublisherHelper#annotateShardSpec} for converting to NumberedShardSpec. Note that, when
+ * the segment lock is used, the Overlord coordinates the segment allocation and this class is never used. See
+ * {@link PartialShardSpec} for that case.
```
Comment (jon-wei) [inline core/src/main/java/org/apache/druid/timeline/partition/BuildingNumberedShardSpec.java:48]: Suggest mentioning here that overwrites with segment locking use `NumberedOverwriteShardSpec`

File: core/src/main/java/org/apache/druid/timeline/partition/PartialShardSpec.java [hunk 2]
```diff
@@ -29,7 +29,10 @@
 
 /**
  * Class to contain all information of a {@link ShardSpec} except for the partition ID.
- * This class is mainly used by the indexing tasks to allocate new segments using the Overlord.
+ * This class is used when the segment allocation is coordinated by the Overlord; when appending segments to an
+ * existing datasource (either streaming ingestion or batch append) or when using segment locking.
+ * The ingestion tasks send all information required for allocating a new segment using this class and the Overlord
+ * determins the partition ID to create a new segment.
```
Comment (jon-wei) [inline core/src/main/java/org/apache/druid/timeline/partition/PartialShardSpec.java:35]: determins -> determines

File: core/src/main/java/org/apache/druid/timeline/partition/PartialShardSpec.java [hunk 3]
```diff
@@ -29,7 +29,10 @@
 
 /**
  * Class to contain all information of a {@link ShardSpec} except for the partition ID.
- * This class is mainly used by the indexing tasks to allocate new segments using the Overlord.
+ * This class is used when the segment allocation is coordinated by the Overlord; when appending segments to an
+ * existing datasource (either streaming ingestion or batch append) or when using segment locking.
```
Comment (jon-wei) [inline core/src/main/java/org/apache/druid/timeline/partition/PartialShardSpec.java:33]: Is this used for all segment locking cases? (is PartialShardSpec also involved in overwrites with segment locking, or is that just OverwriteShardSpec?)

File: core/src/main/java/org/apache/druid/timeline/partition/BuildingNumberedShardSpec.java (hunk 1 above)
Comment (jihoonson) [reply core/src/main/java/org/apache/druid/timeline/partition/BuildingNumberedShardSpec.java]: In segment locking, any implementation of `OverwriteShardSpec` can be used when you overwrite segments. Rephrased the last sentence.

File: core/src/main/java/org/apache/druid/timeline/partition/PartialShardSpec.java (hunk 2 above)
Comment (jihoonson) [reply core/src/main/java/org/apache/druid/timeline/partition/PartialShardSpec.java]: Fixed, thanks.

File: core/src/main/java/org/apache/druid/timeline/partition/PartialShardSpec.java (hunk 3 above)
Comment (jihoonson) [reply core/src/main/java/org/apache/druid/timeline/partition/PartialShardSpec.java]: Yes, it's used all cases when segment lock is used. `PartialShardSpec` is used in the segment allocation protocol when the overlord coordinates the allocation. So, it is involved in 1) all cases with segment locking and 2) appending with time chunk locking. `OverwriteShardSpec` is a special shard spec used with segment locking for the segments overwriting others. Added some more details.

**Comment by jihoonson** [comment]
Tagged "Release Notes" as this PR changes the behavior in loading dynamically-partitioned segments.

**Pull request merged**